DM-44161: Add a representative ConsDB schema.

ktlim opened the pull request.

JeremyMcCormick reviewed (changes requested):
I'd recommend the following changes:

1. Remove `yml/post.sh` as this directory is strictly for YAML files.
2. Recommend to remove repeated `nullable: true` fields, as this is the column default now.
3. Follow consistent table and column naming conventions. If following other schemas this would be capitalized table names and lowercase column names. This is up to you and application specific, but following a consistent convention across the schema would be preferred. 
4. I would change the name of the schema to match the database name, e.g., summit-latiss.yaml -> cdb_latiss.yaml, unless there is a legitimate reason to make it different.

ktlim commented:
- post.sh is gone.
- I thought I had removed the `nullable: true` lines, but I see I missed most of them somehow.
- I really don't want to force people to use capitalized — and especially quoted — table names in queries.  But having them capitalized in the schema browser is nice, and as you say it matches other schemas.  The column names are even more difficult, as they're meant to match the topic names in the EFD.
- I have changed the name of the YAML file.
- I also squashed to a single commit.

JeremyMcCormick commented:
@ktlim There is new validation to Felis being added soon which checks for lengths on types that are considered sized.

I'm seeing a few errors in the schema when I run with this check:

```
ERROR:felis:5 validation errors for Schema
tables.7.columns.1
  Value error, Length must be provided for type 'string' in column '#Exposure_flexdata.key' [type=value_error, input_value={'name': 'key', '@id': '#...iption': 'Name of key.'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.6/v/value_error
tables.7.columns.2
  Value error, Length must be provided for type 'string' in column '#Exposure_flexdata.value' [type=value_error, input_value={'name': 'value', '@id': ... of value as a string.'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.6/v/value_error
tables.8.columns.0
  Value error, Length must be provided for type 'string' in column '#Exposure_flexdata_schema.key' [type=value_error, input_value={'name': 'key', '@id': '#...iption': 'Name of key.'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.6/v/value_error
tables.8.columns.1
  Value error, Length must be provided for type 'string' in column '#Exposure_flexdata_schema.dtype' [type=value_error, input_value={'name': 'dtype', '@id': ...ata type of the value.'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.6/v/value_error
tables.8.columns.2
  Value error, Length must be provided for type 'string' in column '#Exposure_flexdata_schema
.doc' [type=value_error, input_value={'name': 'doc', '@id': '#...de units and IVOA UCD.'}, in
put_type=dict]
    For further information visit https://errors.pydantic.dev/2.6/v/value_error
```

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    length: 32
+    description: Name of the band used to take the exposure where this source was
+      measured. Abstract filter that is not associated with a particular instrument.
+  - name: ra
```
I think we should consider using ObsCore names for columns that have those roles in this schema.  So this one would be `s_ra`.  In that case you can also set `tap:std` to 1.

The standard metadata should also be included: UCD `pos.eq.ra` and units `deg`.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#Exposure.ra"
+    datatype: double
+    description: RA of targeted focal plane center.
+  - name: decl
```
`s_dec`, but even if that's controversial, didn't we agree long ago to move everything from `decl` to `dec`?

ebellm replied:
yes, https://rubinobs.atlassian.net/browse/RFC-863

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#Exposure.decl"
+    datatype: double
+    description: Declination of targeted focal plane center
+  - name: skyRotation
```
Very much needs units.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#Exposure.skyRotation"
+    datatype: double
+    description: Targeted sky rotation angle.
+  - name: azimuth_start
```
Many of the attributes in this section have well-defined UCDs that could be included.  This isn't essential because there isn't current a lot of client behavior tied to them, to be honest, but still should be done in a followup.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#Exposure.airmass"
+    datatype: double
+    description: Airmass of the observed line of sight at the middle of the exposure.
+  - name: expMidpt
```
The `length` of the `timestamp` datatype is a problematic concept, as the modifier in the underlying database type is a precision, not a length.  `64` is unlikely to be a useful value in any event.

gpdf replied:
The same comment applies in several other places.

JeremyMcCormick replied:
Using `postgresql:datatype: TIMESTAMP(6)` might be the best option for now given current limitations in the Felis data model.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    length: 64
+    description: Midpoint time for exposure at the fiducial center of the focal plane
+      array. TAI, accurate to 10ms.
+  - name: expMidptMJD
```
This must have unit (`d`) and a `time.epoch` UCD.  We do have client behaviors tied to that.

gpdf replied:
The same comment applies in several other places.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    datatype: string
+    length: 1024
+    description: Target of the observation.
+  - name: shutterOpenBegin
```
Why are these `timestamp` values and not MJDs?  Is there a substantive benefit?

ktlim replied:
I did not transform them from the original.  Would it be better if I did?

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    datatype: timestamp
+    length: 64
+    description: End of shutter opening.  TAI, accurate to 10ms.
+  - name: airTemp
```
Units required

ktlim replied:
My reading of the VOUnits standard is that this would require a conversion to K, as degC is not allowed.  Is that correct?  Does that make this value less usable, if more correct?

ktlim replied:
Left as `degC` in the description with no official unit.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#Exposure.airTemp"
+    datatype: float
+    description: Outside air temperature.
+  - name: pressure
```
Units required

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#Exposure.humidity"
+    datatype: float
+    description: Outside relative humidity.
+  - name: wind_speed
```
Units required

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    nullable: false
+    description: Detector ID. A detector associated with a particular instrument (not
+      an observation of that detector)
+  - name: xSize
```
Would prefer ObsCore `s_xel1` and `s_xel2` here.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#CcdExposure.ySize"
+    datatype: long
+    description: Number of rows in the image.
+  - name: llcra
```
Do we have code that actually uses the corner positions from this table?  I'm not against having them, but I'm curious about how we use them.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#CcdExposure.lrcra"
+    datatype: double
+    description: RA of lower right corner.
+  - name: lrcdec
```
Also turning all the corners into an `s_region` string makes the table immediately more usable in client software.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#CcdExposure.lrcdec"
+    datatype: double
+    description: Declination of lower right corner.
+  - name: temp_set
```
units required

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#CcdExposure.temp_set"
+    datatype: float
+    description: Temperature setpoint of CCD.
+  - name: ccdtemp
```
units required

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    length: 32
+    description: Name of the band used to take the exposure where this source was
+      measured. Abstract filter that is not associated with a particular instrument.
+  - name: ra
```
`s_ra` preferred.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    "@id": "#Visit.ra"
+    datatype: double
+    description: RA of focal plane center.
+  - name: decl
```
`s_dec` preferred.  Second choice `dec`.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+    description: Name of key.
+  - name: dtype
+    "@id": "#Exposure_flexdata_schema.dtype"
+    datatype: string
```
Please specify what vocabulary is being used for data types here.

Can the VOTable vocabulary be used?
https://ivoa.net/documents/VOTable/20191021/REC-VOTable-1.4-20191021.html#ToC11

ktlim replied:
Specified the vocabulary explicitly.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+  - name: doc
+    "@id": "#Exposure_flexdata_schema.doc"
+    datatype: string
+    description: Documentation string.  Can include units and IVOA UCD.
```
Units and UCDs should be broken out as separate columns.  See the TAP_SCHEMA.columns schema for guidance.
https://ivoa.net/documents/TAP/20190927/REC-TAP-1.1.html#tth_sEc4.3

ktlim replied:
Since this schema is dynamic and intended for internal engineering use only, I don't think it will be translated to TAP Schema.

ktlim replied:
After discussion, conceded that Gregory is correct, as usual.

gpdf reviewed (changes requested):
I have two tiers of comments.

I would like to have column names for standards-compatible columns set appropriately from the start, as I anticipate enormous resistance to changing key column names after this is deployed.

There are also missing annotations throughout.  These don't have to be changed immediately, but for quantities for which the units are not universally standardized, the units really need to be included.  I'm particularly concerned with temperatures, and also with some angles for which there's practice to have them in either radians or degrees.

Finally, searching by time will be greatly assisted by annotating MJD times with UCDs and units.

I have not commented on every individual line that I think requires work, as a lot of this is repetitive.

ktlim commented:
Three overall comments:
* I have no particular stake in the naming of things, but I also didn't see a need for these tables to directly align with ObsCore if that is mostly going to be separate tables (generated by the Butler) and/or a view on these tables.
* Generally the types of columns that come from known places in the headers or EFD topics were aligned with their sources.  However, conversions and transformations are a part of the ConsDB's role, so if science would be better served with a different value, that's fine.
* Flexible metadata is not intended for publication outside the project — it is supposed to be for development and experimentation — so I did not want to demand a higher level of documentation and was hoping not to have to support many datatypes.

gpdf commented:
The argument for using standard metadata for columns whose semantics do in fact align with an existing standard is to produce a better user experience regardless of what client tools are being used.  Increasingly, tools like TOPCAT and PyVO are going to do useful things in response.

> Generally the types of columns that come from known places in the headers or EFD topics were aligned with their sources. However, conversions and transformations are a part of the ConsDB's role, so if science would be better served with a different value, that's fine.

I'm not entirely sure what this is in response to.  I think the `timestamp` types were the only ones we expressed concern about?  Is that about them?

> Flexible metadata is not intended for publication outside the project — it is supposed to be for development and experimentation — so I did not want to demand a higher level of documentation and was hoping not to have to support many datatypes.

That's fine, but I still think if the data are going to be used programmatically we'll need to specify what the datatype vocabulary actually is.  I would be happiest with the VOTable vocabulary, but if you want to specify it as something else (e.g., the numpy type system), we just need to get that stated.  Otherwise people will just do random things.

andy-slac commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,829 @@
+---
```
I expect to see a number of updates for this type of schema. Maybe it's worth thinking about schema versioning and schema upgrades like what we added to APDB recently (https://dmtn-269.lsst.io/).

ktlim replied:
There will be a million updates to this, yes.  I am hoping that @JeremyMcCormick will come up with an Alembic-based way to deal with versioning and upgrades.

ktlim commented:
I've made a number of the requested changes, including changing the names of many columns and adding information, but I'm not sure what to do with `degC` units.

JeremyMcCormick reviewed (approved):
All of the changes I requested in https://github.com/lsst/sdm_schemas/pull/214#pullrequestreview-2037103418 and https://github.com/lsst/sdm_schemas/pull/214#issuecomment-2093348196 have been addressed.

Please wait on the additional approval from @gpdf before merging.

gpdf commented on yml/cdb_latiss.yaml:
```diff
@@ -0,0 +1,359 @@
+    datatype: text
+    nullable: false
+    description: Name of key.
+  - name: dtype
```
This is nullable?

gpdf reviewed (commented):
Looks good now.  Let's see it run!

gpdf reviewed (approved):
(trying again)
